if.1
.\"
.\" Copyright (c) 2004-2011
.\"	Jeffrey Allen Neitzel <jan (at) v6shell (dot) org>.
.\"	All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JEFFREY ALLEN NEITZEL ``AS IS'', AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL JEFFREY ALLEN NEITZEL BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	@(#)$Id$
.\"
.\"	Derived from: Sixth Edition UNIX /usr/man/man1/if.1
.\"
.\" Copyright (C) Caldera International Inc.  2001-2002.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code and documentation must retain the above
.\"    copyright notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\" 4. Neither the name of Caldera International, Inc. nor the names of other
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE LIABLE FOR ANY DIRECT,
.\" INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH IF 1 "@OSH_DATE@" "@OSH_VERSION@" "General Commands"
.SH NAME
if \- conditional command
.SH SYNOPSIS
.B if
[\fIexpression\fR [\fIcommand\fR [\fIarg ...\fR]]]
.SH DESCRIPTION
.I If
evaluates the specified
.IR expression ,
and if its value is true,
returns a zero exit status or executes
the specified
.I command
with the given arguments.
Otherwise,
it returns a non-zero exit status.
When
.I expression
is not specified,
.I if
also returns a non-zero exit status.
.PP
The following primaries are used to construct
.IR expression :
.TP 10
\fB{\fR \fIcommand\fR [\fIarg ...\fR] \fB}\fR
The specified \fIcommand\fR is executed with the
given arguments to obtain its exit status.
A zero status is \fItrue\fR;
a non-zero status is \fIfalse\fR.
.TP
.BI \-d \ file
True if \fIfile\fR exists and is a directory.
.TP
.BI \-e \ file
True if \fIfile\fR exists.
.TP
.BI \-f \ file
True if \fIfile\fR exists and is a regular file.
.TP
.BI \-h \ file
True if \fIfile\fR exists and is a symbolic link.
.TP
.BI \-r \ file
True if \fIfile\fR exists and is readable.
.TP
.BI \-s \ file
True if \fIfile\fR exists and has a size greater than zero bytes.
.TP
.BI \-t \ fildes
True if the file whose file descriptor number is
.I fildes
is open and associated with a terminal device.
.I Fildes
must be a decimal digit (0 \- 9).
.TP
.BI \-w \ file
True if \fIfile\fR exists and is writable.
.TP
.BI \-x \ file
True if \fIfile\fR exists and is executable,
or if \fIfile\fR is a searchable directory.
.TP
.IB file1 \ \-ef \ file2
True if \fIfile1\fR and \fIfile2\fR both exist and refer
to the same file (same device, same inode).
.TP
.IB file1 \ \-nt \ file2
True if \fIfile1\fR and \fIfile2\fR both exist
and last data-modification time of \fIfile1\fR
is newer than that of \fIfile2\fR.
.TP
.IB file1 \ \-ot \ file2
True if \fIfile1\fR and \fIfile2\fR both exist
and last data-modification time of \fIfile1\fR
is older than that of \fIfile2\fR.
.TP
.IB s1 \ = \ s2
True if the strings
.I s1
and
.I s2
are equal.
.TP
.IB s1 \ != \ s2
True if the strings
.I s1
and
.I s2
are not equal.
.PP
These primaries may also be combined
with the following operators:
.TP 10
.BI ! \ expression
unary
.I negation
operator
.TP
.IB expression1 \ \-a \ expression2
binary
.I and
operator
.TP
.IB expression1 \ \-o \ expression2
binary
.I or
operator
.TP
.BI ( \ expression \ )
parentheses for grouping
.PP
.B \-a
has higher precedence than
.BR \-o .
Notice that all of the operators and flags
are separate arguments to
.IR if .
Notice also that parentheses are meaningful
to the shell and must be escaped.
.PP
In addition,
except for the
.B \-h
primary,
symbolic links are followed for all
other \fIfile\fR-related primaries.
.\" primaries that refer to filesystem objects.
.\" primaries that refer to files.
.SH "EXIT STATUS"
The
.I if
command exits with one of the following values:
.TP
0
The expression was true (see below).
.TP
1
The expression was false or was not specified.
.TP
2
An error was detected.
.TP
125
The specified command was found
but did not begin with the proper
magic number or a `#!shell' sequence,
and a valid shell was not specified by
EXECSHELL
with which to execute it.
.TP
126
The specified command was found
but could not be executed.
.TP
127
The specified command was not found.
.PP
If the expression is true and if
.I command
is specified and executed,
the exit status is that of the executed
.IR command .
.SH ENVIRONMENT
Notice that the concept of `user environment'
was not defined in Sixth Edition UNIX.
Thus,
use of the following environment variables
by this port of the conditional command is an enhancement:
.TP
.B EXECSHELL
If set to a non-empty string,
the value of this variable is taken as the
path name of the shell which is invoked to
execute the specified command when it does not
begin with the proper magic number
or a `#!shell' sequence.
.TP
.B PATH
If set to a non-empty string,
the value of this variable is taken as the
sequence of directories which is used to
search for the specified command.
Notice that the
conditional command from Sixth Edition UNIX
always used the equivalent of `.:/bin:/usr/bin',
not PATH.
.SH "SEE ALSO"
goto(1),
osh(1),
sh6(1),
test(1)
.PP
Osh home page:
http://v6shell.org/
.SH COMPATIBILITY
The
.I if
command from Sixth Edition UNIX
does not support the
.BR \-d ,
.BR \-e ,
.BR \-f ,
.BR \-h ,
.BR \-s ,
.BR \-t ,
.BR \-x ,
.BR \-ef ,
.BR \-nt ,
and
.B \-ot
operators.
.PP
In addition to supporting the above operators,
this port also differs from the original in that
the exit status returned varies according to whether
the expression is true or false,
as is the case with
.IR test (1).
.SH HISTORY
An
.I if
command
appeared as
.I /bin/if
in Third Edition UNIX.
.SH LICENSE
See either the LICENSE file which is distributed with
.I osh
or
http://v6shell.org/license/
for full details.
.SH COPYRIGHT
.nf
Copyright (c) 2004-2011
	Jeffrey Allen Neitzel.  All rights reserved.

Copyright (c) 2001-2002
	Caldera International Inc.  All rights reserved.

Copyright (c) 1985, 1989, 1991, 1993
	The Regents of the University of California.  All rights reserved.
.fi


CHANGES
Changes that improve compatibility w/ the Sixth Edition (V6) UNIX
Thompson shell are marked w/ a `C:' in the details below.

-------------------------------------------------------------------------------
[osh-current]: Friday, 2010-05-14:
osh.1:
	* Added [--] to fd2 usage synopsis in "Special commands" subsection.

	* s/Mac/Mac OS X/

-------------------------------------------------------------------------------
[osh-20100430]:
osh.c:
	* Added a new, osh history-file-support feature to save the user's
	  command-line history to the user $h/.osh.history file.  Notice
	  that the history file must already exist as a writable, regular
	  file (or a link to a regular file) when the shell is invoked to
	  save the user's command-line history.  Otherwise, it won't do so.

	  An interactive shell reads each command line from its terminal and
	  writes the words of each one to the history file as a history entry
	  after performing parameter substitution and word splitting.

osh.1:
	* Documented the new, osh history-file-support feature in the new
	  "History file (+)" subsection.

-------------------------------------------------------------------------------
[osh-20100228]:
mkconfig:
	* Added a search comparison that uses which(1) to find login & newgrp
	  pathname(s).  This allows for those cases where a sysadmin may use
	  locally modified versions of these utilities to deal w/ uniquely
	  local needs.  Notice that this is in addition to how login & newgrp
	  pathname(s) were searched for before now.

osh.c:
	* Changed osh globbing so it's handled in main osh process.  This is
	  for user convenience to allow (for example):

		% mkdir some_really_long_directory_name... ; cd some*...

	  ...
	  and globbing for other special built-in commands as well.

-------------------------------------------------------------------------------
[osh-20091218]:
	* Incorporated osh-20091127-p1.patch changes (see p1: below).

INSTALL, Makefile:
	* Made "install-doc" and "install-exp" targets optional so that
	  they're no longer triggered by default "install" target(s).

mkconfig:
	* p1: Defined CONFIG_SUNOS on OpenSolaris (uname -s == SunOS).

goto.c:
	* p1: Fixed problem w/ external goto command on OpenSolaris where
	      file offset set via lseek(2) isn't shared w/ parent process,
	      causing unexpected behavior.

examples/*osh*:
	* Made changes that allow users to use osh to initialize/invoke sh6
	  by doing `[exec] osh [-i | -l] sh6'. This is most convenient way
	  to perform startup actions for sh6 without adding undesirable
	  functionality to sh6 itself.

examples/dot.osh.logout:
	* Fixed `rm -r $d' error on OpenSolaris.

-------------------------------------------------------------------------------
[osh-20091127]:
err.c:
	* Changed setmyname(argv[0]) so that it only ditches an initial
	  "-" character when a binary name is "-" plus at least 1 of any
	  character before end-of-string.  For example:

		% pwd ; ln -s osh - ; ln -s osh -o ; ln -s osh -os ; \
			ln -s osh -osh ; echo -* osh
		/tmp/v6shell
		- -o -os -osh osh

		% which - -o -os -osh osh
		/tmp/v6shell/-
		/tmp/v6shell/-o
		/tmp/v6shell/-os
		/tmp/v6shell/-osh
		/tmp/v6shell/osh

		% - file arg1 ; -o file arg1 ; -os file arg1 ; \
			-osh file arg1 ; osh file arg1
		-: file: cannot open
		o: file: cannot open
		os: file: cannot open
		osh: file: cannot open
		osh: file: cannot open

-------------------------------------------------------------------------------
[osh-20091122]:
defs.h:
	* Defined SBUFMM(m), small buffer max multiplier, for calculating
	  small buffer sizes (e.g., DOLMAX, FMTMAX, LABELMAX).

err.h:
	* Changed ERR_ALTOOLONG to ERR_E2BIG, and added FMT4S.

pexec.[hc]:
	* Changed err_pexec() function prototype.

	* Changed it to print ERR_E2BIG diagnostic when errno == E2BIG.

	* Changed exit status to 126 for such "cannot execute" conditions.

*.c:
	* Changed all err_pexec() calls to match new function prototype
	  where appropriate.

osh.c:
	* Changed main child process execute function from pexec()
	  to err_pexec() for external commands.

	* Added getmyname() program name prefix (e.g., "osh: ")
	  to osh diagnostics.

v.c:
	* Added "The Regents of the University of California" copyright notice
	  for NetBSD and OpenBSD code included in the osh project.

[fgio][dfos]*.1:
	* Added COPYRIGHT section headings, and added copyright notices.

-------------------------------------------------------------------------------
[osh-20091029]:
	* Remove rcsid.h file; stop adding rcsid strings for all
	  sources to compiled binaries.

	* Re-add example rc (init and logout) files as examples/* .

	* Add $h/.project file as PROJECT for quick project info.

Makefile:
	* Add default DOCDIR as $(PREFIX)/share/doc/$(OSH_VERSION);
	  use it for [ACILNPR]* package documentation and examples/* .

	* Change default MANDIR to $(PREFIX)/share/man/man1 .

v.c:
	* Include OSH_VERSION from Makefile and project copyright
	  notices in compiled binaries.

fd2.c, if.c, pexec.[hc], util.c:
	* Add err_pexec() error-handling wrapper around pexec() for
	  shell utilities that execute external commands.

osh.c:
	* Wrap WCOREDUMP(s) w/ #ifdef WCOREDUMP ... #endif .

osh.1:
	* Change example in "I/O redirection" subsection that could
	  confuse new users.

-------------------------------------------------------------------------------
[osh-20090527]:
	* Moved inclusion of required kernel include files (e.g., <sys/*.h>)
	  and /usr/include files from *.c to "defs.h" where appropriate.

	* Moved files:	Moved manual pages from *.1.in to *.1 .

	* New files:	err.h, err.c  (see err.[hc] description below)

Makefile:
	* Changed "-O2" to "-Os" in CFLAGS to optimize for size.

	* Changed how the manual pages are generated and installed.

README:
	* Added Debian 5.0.1 (lenny), NetBSD 5.0, OpenBSD 4.5, and
	  OpenSolaris 2008.11 to the reported system list; the osh
	  package behaves as expected on each system.

err.[hc]:

if.c
/*
 * if.c - a port of the Sixth Edition (V6) UNIX conditional command
 */
/*-
 * Copyright (c) 2004-2011
 *	Jeffrey Allen Neitzel <jan (at) v6shell (dot) org>.
 *	All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY JEFFREY ALLEN NEITZEL ``AS IS'', AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL JEFFREY ALLEN NEITZEL BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *	@(#)$Id$
 */
/*
 *	Derived from:
 *		- Sixth Edition UNIX	/usr/source/s1/if.c
 *		- Seventh Edition UNIX	/usr/src/cmd/test.c
 */
/*-
 * Copyright (C) Caldera International Inc.  2001-2002.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code and documentation must retain the above
 *    copyright notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed or owned by Caldera
 *      International, Inc.
 * 4. Neither the name of Caldera International, Inc. nor the names of other
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
 * INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE LIABLE FOR ANY DIRECT,
 * INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "defs.h"
#include "err.h"
#include "pexec.h"
#include "sasignal.h"

static	int	ac;
static	int	ap;
static	char	**av;
static	uid_t	ifeuid;

/*@noreturn@*/
static	void	doex(bool);
static	bool	e1(void);
static	bool	e2(void);
static	bool	e3(void);
static	bool	equal(/*@null@*/ const char *, /*@null@*/ const char *);
static	bool	expr(void);
static	bool	ifaccess(/*@null@*/ const char *, int);
static	bool	ifstat1(/*@null@*/ const char *, mode_t);
static	bool	ifstat2(/*@null@*/ const char *, /*@null@*/ const char *, int);
/*@null@*/
static	char	*nxtarg(bool);

/*
 * NAME
 *	if - conditional command
 *
 * SYNOPSIS
 *	if [expression [command [arg ...]]]
 *
 * DESCRIPTION
 *	See the if(1) manual page for full details.
 */
int
main(int argc, char **argv)
{
	bool re;		/* return value of expr() */

	setmyerrexit(&ut_errexit);
	setmyname(argv[0]);
	setmypid(getpid());

	ifeuid = geteuid();

	/*
	 * Set-ID execution is not supported.
	 */
	if (ifeuid != getuid() || getegid() != getgid())
		err(FC_ERR, FMT2S, getmyname(), ERR_SETID);

	(void)sasignal(SIGCHLD, SIG_DFL);

	if (argc > 1) {
		ac = argc;
		av = argv;
		ap = 1;
		re = expr();
		if (re && ap < ac)
			doex(!FORKED);
	} else
		re = false;

	return re ? SH_TRUE : SH_FALSE;
}

/*
 * Evaluate the expression.
 * Return true (1) or false (0).
 */
static bool
expr(void)
{
	bool re;

	re = e1();
	if (equal(nxtarg(RETERR), "-o"))
		return re | expr();
	ap--;
	return re;
}

static bool
e1(void)
{
	bool re;

	re = e2();
	if (equal(nxtarg(RETERR), "-a"))
		return re & e1();
	ap--;
	return re;
}

static bool
e2(void)
{

	if (equal(nxtarg(RETERR), "!"))
		return !e3();
	ap--;
	return e3();
}

static bool
e3(void)
{
	bool re;
	pid_t cpid, tpid;
	int cstat, d;
	char *a, *b;

	if ((a = nxtarg(RETERR)) == NULL)
		err(FC_ERR, FMT3S, getmyname(), av[ap - 2], ERR_EXPR);

	/*
	 * Deal w/ parentheses for grouping.
	 */
	if (equal(a, "(")) {
		re = expr();
		if (!equal(nxtarg(RETERR), ")"))
			err(FC_ERR, FMT3S, getmyname(), a, ERR_PAREN);
		return re;
	}

	/*
	 * Execute { command [arg ...] } to obtain its exit status.
	 */
	if (equal(a, "{")) {
		if ((cpid = fork()) == -1)
			err(FC_ERR, FMT2S, getmyname(), ERR_FORK);
		if (cpid == 0)
			/**** Child! ****/
			doex(FORKED);
		else {
			/**** Parent! ****/
			tpid = wait(&cstat);
			while ((a = nxtarg(RETERR)) != NULL && !equal(a, "}"))
				;	/* nothing */
			if (a == NULL)
				ap--;
			return (tpid == cpid && cstat == 0) ? true : false;
		}
	}

	/*
	 * file existence/permission tests
	 */
	if (equal(a, "-e"))
		return ifaccess(nxtarg(!RETERR), F_OK);
	if (equal(a, "-r"))
		return ifaccess(nxtarg(!RETERR), R_OK);
	if (equal(a, "-w"))
		return ifaccess(nxtarg(!RETERR), W_OK);
	if (equal(a, "-x"))
		return ifaccess(nxtarg(!RETERR), X_OK);

	/*
	 * file existence/type tests
	 */
	if (equal(a, "-d"))
		return ifstat1(nxtarg(!RETERR), S_IFDIR);
	if (equal(a, "-f"))
		return ifstat1(nxtarg(!RETERR), S_IFREG);
	if (equal(a, "-h"))
		return ifstat1(nxtarg(!RETERR), S_IFLNK);
	if (equal(a, "-s"))
		return ifstat1(nxtarg(!RETERR), F_GZ);
	if (equal(a, "-t")) {
		/* Does the descriptor refer to a terminal device? */
		b = nxtarg(RETERR);
		if (b == NULL || *b == EOS)
			err(FC_ERR, FMT3S, getmyname(), a, ERR_DIGIT);
		if (*b >= '0' && *b <= '9' && *(b + 1) == EOS) {
			d = *b - '0';
			if (IS_DIGIT(d, *b))
				return isatty(d) != 0;
		}
		err(FC_ERR, FMT3S, getmyname(), b, ERR_NOTDIGIT);
	}

	/*
	 * binary comparisons
	 */
	if ((b = nxtarg(RETERR)) == NULL)
		err(FC_ERR, FMT3S, getmyname(), a, ERR_OPERATOR);
	if (equal(b,  "="))
		return  equal(a, nxtarg(!RETERR));
	if (equal(b, "!="))
		return !equal(a, nxtarg(!RETERR));
	if (equal(b, "-ot"))
		return ifstat2(a, nxtarg(!RETERR), F_OT);
	if (equal(b, "-nt"))
		return ifstat2(a, nxtarg(!RETERR), F_NT);
	if (equal(b, "-ef"))
		return ifstat2(a, nxtarg(!RETERR), F_EF);
	err(FC_ERR, FMT3S, getmyname(), b, ERR_OPUNKNOWN);
	/*NOTREACHED*/
	return false;
}

static void
doex(bool forked)
{
	char **xap, **xav;

	if (ap < 2 || ap > ac)	/* should never (but can) be true */
		err(FC_ERR, FMT2S, getmyname(), ERR_AVIINVAL);

	xav = xap = &av[ap];
	while (*xap != NULL) {
		if (forked && equal(*xap, "}"))
			break;
		xap++;
	}
	if (forked && xap - xav > 0 && !equal(*xap, "}"))
		err(FC_ERR, FMT3S, getmyname(), av[ap - 1], ERR_BRACE);
	*xap = NULL;
	if (xav[0] == NULL) {
		if (forked)
			err(FC_ERR, FMT3S, getmyname(),av[ap - 1],ERR_COMMAND);
		else
			/* Currently unreachable; do not remove. */
			err(FC_ERR, FMT2S, getmyname(), ERR_COMMAND);
	}

	/* Invoke a special "exit" utility in this case. */
	if (equal(xav[0], "exit")) {
		(void)lseek(FD0, (off_t)0, SEEK_END);
		EXIT(SH_TRUE);
	}

	(void)err_pexec(xav[0], xav);
}

/*
 * Check access permission for file according to mode while
 * dealing w/ special cases for the superuser and `-x':
 *	- Always grant search access on directories.
 *	- If not a directory, require at least one execute bit.
 * Return true  (1) if access is granted.
 * Return false (0) if access is denied.
 */
static bool
ifaccess(const char *file, int mode)
{
	struct stat sb;
	bool ra;

	if (file == NULL || *file == EOS)
		return false;

	ra = access(file, mode) == 0;

	if (ra && mode == X_OK && ifeuid == 0) {
		if (stat(file, &sb) < 0)
			ra = false;
		else if (S_ISDIR(sb.st_mode))
			ra = true;
		else
			ra = (sb.st_mode & (S_IXUSR|S_IXGRP|S_IXOTH)) != 0;
	}

	return ra;
}

static bool
ifstat1(const char *file, mode_t type)
{
	struct stat sb;
	bool rs;

	if (file == NULL || *file == EOS)
		return false;

	if (type == S_IFLNK) {
		if (lstat(file, &sb) < 0)
			rs = false;
		else
			rs = (sb.st_mode & S_IFMT) == type;
	} else if (stat(file, &sb) < 0)
		rs = false;
	else if (type == S_IFDIR || type == S_IFREG)
		rs = (sb.st_mode & S_IFMT) == type;
	else if (type == F_GZ)
		rs = sb.st_size > (off_t)0;
	else
		rs = false;

	return rs;
}

static bool
ifstat2(const char *file1, const char *file2, int act)
{
	struct stat sb1, sb2;
	bool rs;

	if (file1 == NULL || *file1 == EOS)
		return false;
	if (file2 == NULL || *file2 == EOS)
		return false;

	if (stat(file1, &sb1) < 0)
		return false;
	if (stat(file2, &sb2) < 0)
		return false;

	if (act == F_OT)
		rs = sb1.st_mtime < sb2.st_mtime;
	else if (act == F_NT)
		rs = sb1.st_mtime > sb2.st_mtime;
	else if (act == F_EF)
		rs = sb1.st_dev == sb2.st_dev && sb1.st_ino == sb2.st_ino;
	else
		rs = false;

	return rs;
}

static char *
nxtarg(bool reterr)
{
	char *nap;

	if (ap < 1 || ap > ac)	/* should never (but can) be true */
		err(FC_ERR, FMT2S, getmyname(), ERR_AVIINVAL);

	if (ap == ac) {
		if (reterr) {
			ap++;
			return NULL;
		}
		err(FC_ERR, FMT3S, getmyname(), av[ap - 1], ERR_ARGUMENT);
	}
	nap = av[ap];
	ap++;
	return nap;
}

static bool
equal(const char *a, const char *b)
{

	if (a == NULL || b == NULL)
		return false;
	return EQUAL(a, b);
}
